prime-router/docs/proposals/0013-batch-retry.md

```diff
@@ -0,0 +1,71 @@
+# Batch Process Update / Retry
+The batching process could use more robust retry functionality and there is an opportunity for additional improvements
+to the functionality as a whole to address scalability and performance concerns.
+
+Associated Ticket - #2902
+
+### Current Batch Functionality  
+Each incoming message puts an entry into the 'batch' queue per receiver that message is going to, using the hidden
+'next action at' part of the queue and selecting the next 'batch bucket' for the associated receiver.
+
+When that time comes all queue messages for the receiver in question are added to the queue and picked up by individual 
+instances of the batch function (one per message on the queue). 
+
+Each of these functions then race to select and lock 
+all records from the Task table that are for the receiver in question. The function instances that fail to get any
+records die.
+
+If a batch function fails to succeed it will be lost in the ether and require manual resetting to pick up and batch
+included records.
+
+### Proposed New Functionality
+We can improve this situation and add retry/scalability by adding a cron job and altering the way we use the batch
+queue. Instead of adding each message-per-receiver-per-bucket to the queue at ingestion a cron job could run at
+the minimum granularity for batching (1 minute), determine which receivers need to be batched at that time, and put a 
+message on the queue to be handled immediately by the batchFunction. This is a minimal-change approach that leaves the
+Task table management as it is but severely reduces our performance bottlenecks and race conditions. Once done, the
+single run of the batch function for that receiver will pull all outstanding records up to [receiver limit] and batch 
+them. If there are more records than [receiver limit] it will continue to follow the batching process until all records
+have been batched.  We will need to take a snapshot of the id of the 'last record' at start of batch time or this could
+be an infinite batching process if records continue to trickle in.
```

Comment:
```suggestion
them. 

If there are more records than [receiver-limit] it will place
 `(count(reports-past-due-for-batching-for-one-receiver) / [receiver-limit] )`
 messages on the queue for that receiver, and depend on Azure to scale to meet that demand, as well as on Azure's queue throttling settings (in `host.json`) to avoid swamping the system.

This will cover situations where, for example, Batch has been down for some time and Batch work is horrendously backed up.
```

@clediggins-usds:    @whytheplatypus and I had a good discussion about this on Friday, and this seemed like a good approach to this problem.